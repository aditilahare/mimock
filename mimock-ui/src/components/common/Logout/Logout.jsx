import React, { useState, useEffect } from 'react';
import { Navigate } from 'react-router';
import { Cookies } from 'react-cookie';
import { useRecoilState } from 'recoil';
import showFullPageLoader from 'atoms/showFullPageLoader';
import { routes } from 'constants/routes';
import { globalConstants } from 'constants/globalConstants';

export default function Logout() {
	const [, setOpenFullPageLoader] = useRecoilState(showFullPageLoader);
	const [clearingSession, setClearingSession] = useState(true);

	useEffect(() => {
		setOpenFullPageLoader(true);

		const cookies = new Cookies();
		cookies.remove(globalConstants.XSRF_COOKIE_NAME, {
			path: '/',
		});
		cookies.remove(globalConstants.AUTH_TOKEN_COOKIE_NAME, {
			path: '/',
		});

		setClearingSession(false);
		setOpenFullPageLoader(false);
	}, []);

	return (
		<If condition={!clearingSession}>
			<Navigate to={routes.root} replace={true} />
		</If>
	);
}
