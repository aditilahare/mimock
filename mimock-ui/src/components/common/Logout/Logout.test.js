import React from 'react';
import { render } from '@testing-library/react';
import Logout from './Logout';

let mockedRecoilFn;
jest.mock('recoil', () => {
	mockedRecoilFn = jest.fn();
	return {
		atom: jest.fn(),
		useRecoilState: jest.fn(() => [true, mockedRecoilFn]),
	};
});

let mockedUseState;
jest.mock('react', () => {
	mockedUseState = jest.fn();
	return {
		...jest.requireActual('react'),
		useState: mockedUseState.mockImplementation(() => [true, jest.fn()]),
	};
});

jest.mock('react-router', () => ({
	...jest.requireActual('react-router'),
	// eslint-disable-next-line react/prop-types
	Navigate: ({ to, replace }) => {
		return (
			<div data-testid='react-router-navigator'>
				<div data-testid='navigate-to'>{to}</div>
				{replace && <div data-testid='navigate-replace'>true</div>}
			</div>
		);
	},
}));

describe('Logout', () => {
	it('should show full page loader when session cleanup is in progress', async () => {
		const { queryByTestId } = await render(<Logout />);

		expect(mockedRecoilFn).toHaveBeenCalledTimes(2);
		expect(mockedRecoilFn).nthCalledWith(1, true);
		expect(mockedRecoilFn).nthCalledWith(2, false);

		expect(queryByTestId('navigate-to')).not.toBeInTheDocument();

		expect(document.body).toMatchSnapshot();
	});

	it('should navigate to login page after clearing session', async () => {
		mockedUseState.mockImplementation(() => [false, jest.fn()]);

		const { container, getByTestId } = await render(<Logout />);

		expect(mockedRecoilFn).toHaveBeenCalledTimes(2);
		expect(mockedRecoilFn).nthCalledWith(1, true);
		expect(mockedRecoilFn).nthCalledWith(2, false);

		expect(getByTestId('react-router-navigator')).toBeInTheDocument();
		expect(getByTestId('navigate-to')).toBeInTheDocument();
		expect(getByTestId('navigate-to').textContent).toBe('/mimock-ui');
		expect(getByTestId('navigate-replace')).toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});
});
