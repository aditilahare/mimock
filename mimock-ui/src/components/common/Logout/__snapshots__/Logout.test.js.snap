// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout should navigate to login page after clearing session 1`] = `
<div>
  <div
    data-testid="react-router-navigator"
  >
    <div
      data-testid="navigate-to"
    >
      /mimock-ui
    </div>
    <div
      data-testid="navigate-replace"
    >
      true
    </div>
  </div>
</div>
`;

exports[`Logout should show full page loader when session cleanup is in progress 1`] = `
<body>
  <div />
</body>
`;
